src/zimdump.cpp

```diff
@@ -385,132 +367,143 @@ void ZimDumper::dumpFiles(const std::string& directory, bool symlinkdump)
-                   "  " << argv[0] << " -f Auto -l -i -v wikipedia.zim\n"
-                   "  " << argv[0] << " -o 123159 -l -i wikipedia.zim\n"
-                 << std::flush;
-      return -1;
+        article = app.getArticleByTitle(args["--idx"].asLong());
+    } else if(args["--url"])
+    {
+        std::string nspace = "A";
+        if (args["--ns"])
+            nspace = args["--ns"].asString();
+        article = app.getArticleByUrl(nspace + "/" + args["--url"].asString());
     }
+    app.dumpArticle(article);
+}
 
-    // initalize app
-    ZimDumper app(argv[1], titleSort);
-    app.setVerbose(verbose);
-
-    // global info
-    if (fileinfo)
-      app.printInfo();
+void subcmdList(ZimDumper &app, std::map<std::string, docopt::value> &args)
+{
+    bool idx = false;
+    bool url = false;
+    bool details = args["--details"].asBool();
 
-    // namespace info
-    if (nsinfo.isSet())
-      app.printNsInfo(nsinfo);
+    if (args["--idx"]) idx = true;
+    if (args["--url"]) url = true;
 
-    // locate article
-    if (indexOffset.isSet())
-      app.locateArticle(indexOffset);
-    else if (find.isSet())
-      app.findArticle(ns, find, titleSort);
-    else if (url.isSet())
-      app.findArticleByUrl(std::string(url));
+    if (idx || url)
+    {
+        zim::Article article;
+        if (idx)
+        {
+            article = app.getArticleByTitle(args["--idx"].asLong());
+        } else if(url)
+        {
+            std::string nspace = "A";
+            if (args["--ns"])
+                nspace = args["--ns"].asString();
+            article = app.getArticleByUrl(nspace + "/" + args["--url"].asString());
+        }
 
-    // dump files
-    if (dumpAll.isSet()) {
-#ifdef _WIN32
-      app.dumpFiles(dumpAll.getValue(), false);
-#else
-      app.dumpFiles(dumpAll.getValue(), redirectSymlink);
-#endif
+        app.listArticle(article, details);
     }
+    else
+    {
+        app.listArticles(details, details);
+    }
+}
 
-    // print requested info
-    if (data)
-      app.dumpArticle();
-    else if (page)
-      app.printPage();
-    else if (list || tableList)
-      app.listArticles(info, tableList, extra);
-    else if (info)
-      app.listArticle(extra);
-
-    if (verifyChecksum)
-      app.verifyChecksum();
-  }
-  catch (const std::exception& e)
-  {
-    std::cerr << e.what() << std::endl;
-    return -2;
-  }
-  return 0;
+int main(int argc, char* argv[])
+{
+    std::map<std::string, docopt::value> args
+        = docopt::docopt(USAGE,
+                         { argv + 1, argv + argc },
+                         true,
+                         "zimdump 1.0");
+
+    try {
+        ZimDumper app(args["<file>"].asString());
+
+        std::unordered_map<std::string, std::function<void(ZimDumper&, decltype(args)&)>> dispatchtable = {
+            {"info",            subcmdInfo },
+            {"dump",            subcmdDump },
+            {"list",            subcmdList }
```

Comment:
Is it missing the "ls" subcommand ? (or docopt detect it is a alias ?)

Reply:
Docopt doesnt seem to detect. I added it